### ISSUE TITLE
lora+极小训练集微调glm4v，训练集的loss降到0.2就降不下去了，是否证明无法在少量样本上拟合？

### ISSUE THREAD

**Issue by tomsonsgs** [THUDM/GLM-4#342]
### System Info / 系統信息

ubuntu python3.8

### Information / 问题信息

- [ ] The official example scripts / 官方的示例脚本
- [ ] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

1，取极少训练集，比如2条
2，基于源代码提供的lora设置
3，训练多轮但loss降低到0.2就降低不下去了

### Expected behavior / 期待表现

对于仅仅2条训练集，是否应该可以做到可全拟合训练集，loss可以降低到0才符合预期？

**Comment by zRzRzRzRzRzRzR** [comment]
样本一般都是按千的单位算的吧，你这个极少是多少呢，一般来说1000以上的数据集才有比较能用的效果

**Comment by tomsonsgs** [comment]
目前只是用2个样本验证模型的训练流程，一般来说两个样本train loss应该可以拟合到0左右，我用swift跑这个没问题，train loss能接近到0，用这个git里面的代码就不行 @zRzRzRzRzRzRzR

**Comment by tomsonsgs** [comment]
目前发现grad norm不太正常，下面是部分日志：
  Batch size = 1
{'eval_loss': 3.60791015625, 'eval_runtime': 8.8325, 'eval_samples_per_second': 2.038, 'eval_steps_per_second': 1.019, 'epoch': 70.0}                                                                              
{'loss': 0.2537, 'grad_norm': 32.75, 'learning_rate': 0.0004866666666666667, 'epoch': 80.0}                                                                                                                        
  3%|████▌                                                                                                                                                                     | 80/3000 [13:02<7:21:14,  9.07s/it]***** Running Evaluation *****
  Num examples = 18
  Batch size = 1
{'eval_loss': 5.074381351470947, 'eval_runtime': 8.9442, 'eval_samples_per_second': 2.012, 'eval_steps_per_second': 1.006, 'epoch': 80.0}                                                                          
{'loss': 0.4575, 'grad_norm': 9.8125, 'learning_rate': 0.00048499999999999997, 'epoch': 90.0}                                                                                                                      
  3%|█████                                                                                                                                                                     | 90/3000 [14:42<7:19:40,  9.07s/it]***** Running Evaluation *****
  Num examples = 18
  Batch size = 1
{'eval_loss': 5.586859703063965, 'eval_runtime': 8.8845, 'eval_samples_per_second': 2.026, 'eval_steps_per_second': 1.013, 'epoch': 90.0}                                                                          
{'loss': 0.2968, 'grad_norm': 0.734375, 'learning_rate': 0.00048333333333333334, 'epoch': 100.0}                                                                                                                   
  3%|█████▋                                                                                                                                                                   | 100/3000 [16:20<7:16:28,  9.03s/it]***** Running Evaluation *****
  Num examples = 18
  Batch size = 1
{'eval_loss': 4.017523765563965, 'eval_runtime': 8.8962, 'eval_samples_per_second': 2.023, 'eval_steps_per_second': 1.012, 'epoch': 100.0}                                                                         
{'loss': 0.2526, 'grad_norm': 11.0625, 'learning_rate': 0.0004816666666666667, 'epoch': 110.0}                                                                                                                     
  4%|██████▏                                                                                                                                                                  | 110/3000 [17:58<7:14:09,  9.01s/it]***** Running Evaluation *****
  Num examples = 18
  Batch size = 1
{'eval_loss': 4.80528450012207, 'eval_runtime': 8.9472, 'eval_samples_per_second': 2.012, 'eval_steps_per_second': 1.006, 'epoch': 110.0}                                                                          
{'loss': 0.296, 'grad_norm': 0.01025390625, 'learning_rate': 0.00048, 'epoch': 120.0}                                                                                                                              
  4%|██████▊                                                                                                                                                                  | 120/3000 [19:38<7:16:40,  9.10s/it]***** Running Evaluation *****
  Num examples = 18
  Batch size = 1
{'eval_loss': 4.7001953125, 'eval_runtime': 8.9171, 'eval_samples_per_second': 2.019, 'eval_steps_per_second': 1.009, 'epoch': 120.0}                                                                              
{'loss': 0.4734, 'grad_norm': 36.0, 'learning_rate': 0.0004783333333333333, 'epoch': 130.0}                                                                                                                        
  4%|███████▎                                                                                                                                                                 | 130/3000 [21:16<7:11:36,  9.02s/it]***** Running Evaluation *****
  Num examples = 18
  Batch size = 1
{'eval_loss': 3.399305582046509, 'eval_runtime': 8.9092, 'eval_samples_per_second': 2.02, 'eval_steps_per_second': 1.01, 'epoch': 130.0}                                                                           
{'loss': 0.2724, 'grad_norm': 18.625, 'learning_rate': 0.0004766666666666667, 'epoch': 140.0}                                                                                                                      
  5%|███████▉                                                                                                                                                                 | 140/3000 [22:55<7:09:55,  9.02s/it]***** Running Evaluation *****
  Num examples = 18
  Batch size = 1
{'eval_loss': 4.966959476470947, 'eval_runtime': 8.8692, 'eval_samples_per_second': 2.029, 'eval_steps_per_second': 1.015, 'epoch': 140.0}                                                                         
{'loss': 0.3061, 'grad_norm': 11.75,

**Comment by tomsonsgs** [comment]
已解决，原始代码里<|assistant|>的token也被预测，但这个token无法被训练拟合导致一直有最低的loss存在，改变下数据准备代码即可

**Comment by lileilai** [comment]
> 已解决，原始代码里<|assistant|>的token也被预测，但这个token无法被训练拟合导致一直有最低的loss存在，改变下数据准备代码即可

我碰到了同样的情况，请问是应该修改哪里的代码呢？谢谢！

**Comment by cxfzzj** [comment]
@tomsonsgs 你好，请问这个问题怎么解决的呢？

已解决，原始代码里<|assistant|>的token也被预测，但这个token无法被训练拟合导致一直有最低的loss存在，改变下数据准备代码即可

**Comment by sixsixcoder** [comment]
> @tomsonsgs 你好，请问这个问题怎么解决的呢？
> 
> 已解决，原始代码里<|assistant|>的token也被预测，但这个token无法被训练拟合导致一直有最低的loss存在，改变下数据准备代码即可

请问你是如何修改的代码，可以分享一下经验吗
